### PR TITLE
ref: Refactor query execution into `NativeDriverReader` abstraction

### PR DESCRIPTION
**Pull request by tkaemming**
This takes a similar approach as getsentry/snuba#357 did for the `NativeDriverBatchWriter`, which seemed to work out pretty well.

See SNS-226.

### CODE REVIEW

File: snuba/reader.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+import re
+from abc import ABC, abstractmethod
+from datetime import datetime
+from typing import TYPE_CHECKING, Any, Mapping, Optional, Sequence
+
+from dateutil.tz import tz
+
+
+if TYPE_CHECKING:
+    from mypy_extensions import TypedDict
```
Comment (tkaemming) [inline snuba/reader.py]: These annotations aren't evaluated at runtime, and I didn't want to make `mypy` a hard dependency quite yet, so this is only evaluated when [`TYPE_CHECKING`](https://docs.python.org/3/library/typing.html#typing.TYPE_CHECKING).

File: snuba/reader.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+class NativeDriverReader(Reader):
+
+    DATE_TYPE_RE = re.compile(r"(Nullable\()?Date\b")
+    DATETIME_TYPE_RE = re.compile(r"(Nullable\()?DateTime\b")
+
+    def __init__(self, client):
+        self.__client = client
+
+    def __transform_result(self, result, with_totals: bool) -> Result:
```
Comment (tkaemming) [inline snuba/reader.py]: This is mostly a direct copy of `scrub_ch_data`, with the exception that the `with_totals` handling at the end is brought over from `raw_query`.

File: snuba/reader.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+            return {"data": data, "meta": meta, "totals": totals}
+        else:
+            return {"data": data, "meta": meta}
+
+    def execute(
+        self,
+        query: str,
+        settings: Optional[Mapping[str, str]] = None,
+        query_id: Optional[str] = None,
+        with_totals: bool = False,
```
Comment (tkaemming) [inline snuba/reader.py]: The HTTP driver will ignore this parameter, and it will eventually be deprecated.

File: snuba/reader.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+            elif self.DATE_TYPE_RE.match(col["type"]):
+                for d in data:
+                    dt = datetime(*(d[col["name"]].timetuple()[:6])).replace(
+                        tzinfo=tz.tzutc()
+                    )
+                    d[col["name"]] = dt.isoformat()
+
+        if with_totals:
+            assert len(data) > 0
+            totals = data.pop(-1)
```
Comment (tkaemming) [inline snuba/reader.py]: This used to be `data, totals = data[:-1], data[-1]`, this should be a bit more efficient.

File: snuba/reader.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+    Column = TypedDict("Column", {"name": str, "type": str})
+    Result = TypedDict(
+        "Result",
+        {
+            "meta": Sequence[Column],
+            "data": Sequence[Mapping[str, Any]],
+            "totals": Mapping[str, Any],
+        },
+        total=False,
```
Comment (fpacifici) [inline snuba/reader.py]: Which key can be missing ?

**Review by fpacifici — COMMENTED** [review]
seems good to me. Just a few clarifications.

File: snuba/reader.py (hunk 5 above)
Comment (tkaemming) [reply snuba/reader.py]: The `totals` row is only included if explicitly asked for by the request. The other two are required. I can [split this up and make `meta` and `data` required and `totals` optional](https://mypy.readthedocs.io/en/latest/more_types.html#mixing-required-and-non-required-items) if you'd like. (In all honesty, I was just lazy and figured this code would be short-lived — see the comment above.)

File: snuba/reader.py (hunk 2 above)
Comment (tkaemming) [reply snuba/reader.py]: I guess it'd be `Tuple[Sequence[Tuple[Any]], Sequence[Tuple[str, str]]]`, per https://github.com/mymarilyn/clickhouse-driver/blob/464d232a7e4caf75608e6507175e0bc6ac813f12/clickhouse_driver/client.py#L185-L189

File: snuba/reader.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+from typing import TYPE_CHECKING, Any, Mapping, Optional, Sequence
+
+from dateutil.tz import tz
+
+
+if TYPE_CHECKING:
+    from mypy_extensions import TypedDict
+
+    Column = TypedDict("Column", {"name": str, "type": str})
+    Result = TypedDict(
```
[the remark this answers is not in the document]
Comment (fpacifici) [reply snuba/reader.py]: Sorry, let me rephrase:
It is not immutability that affects multiple implementations. Immutability is good per se. Dataclasses may make it easier to manage multiple implementation since we can have inheritance (I don't htink we can extend a Typed dictionary)

File: snuba/reader.py (hunk 2 above)
Comment (fpacifici) [reply snuba/reader.py]: Did you leave it out intentionally because it comes from non typed code anyway ?

**Review by fpacifici — APPROVED** [review]

File: snuba/reader.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (tkaemming) [reply snuba/reader.py]: Ah, yeah — I agree with this.

File: snuba/reader.py (hunk 2 above)
Comment (tkaemming) [reply snuba/reader.py]: Yeah, I figured this was probably not super useful right now, though if/when we start using mypy with CI it might be more beneficial (if this code is still around.)
